Review (#104)
zarr-developers/zarr-specs: docs/protocol/core/v3.0.rst

We solicit feedback on the following area during the RFC period of this first
draft.

>  - https://github.com/zarr-developers/zarr-specs/issues/72 to potentially split

Reviewer: ```suggestion
 - Should core metadata and user attributes be stored together or separate documents? ([GH72](https://github.com/zarr-developers/zarr-specs/issues/72)) 
```

   large metadata documents.
 - extensions and ``must_understand = True`` might be too restrictive. Work a

>    draft implementation with extensions and

Reviewer: I don't understand the syntax "work a draft implementation"? Do you mean "develop a draft implementation"?

Author: My change is was just on the form. @Carreau might have written it?

>    see how far we can go. List of extensions to implement:

Reviewer: ```suggestion
   see how far we can go. Possible list of extensions to implement:
```

encoded for storage.

Only the top level metadata document ``zarr.json`` is guaranteed to be of JSON
type, and can be used to define other formats for array-level and group-level
metadata documents. In the case where non-JSON metadata documents are used in a
Zarr hierarchy, the following sections on group and array level metadata are

> non-normative; but other metadata formats are expected to define some

Reviewer: ```suggestion
non-normative, but other metadata formats are expected to define some
```

**Store element values in an array**

    To store element in an array at path `P` and coordinate (`i`, `j`,
    ...), perform ``set(data_key(P, i, j, ...), value)``, where
    `value` is the serialisation of the corresponding chunk, encoded
    according to the information in the array metadata stored under
    the key ``array_meta_key(P)``.

**Retrieve element values in an array**

    To retrieve element in an array at path `P` and coordinate (`i`,
    `j`, ...), perform ``get(data_key(P, i, j, ...))``. The returned

>     value is the serialisation of the corresponding chunk, encoded

Author: It was probably due to a copy/past from above, since `get` *returns* the value.

**Discover all nodes in a hierarchy**

    To discover all nodes in a hierarchy, one can call

>     ``list_prefix("meta/root/")``. All keys represent either explicit group or

Author: @alimanfoo `list` doesn't accept any parameter, so I guess we should use `list_prefix`, and I guess that `/root` was missing?

Reviewer 2: I think maybe just `list()` ? 

```
``list`` - Retrieve all `keys` in the store.

    | Parameters: none
    | Output: set of `keys`
```

Or do we get rid of list and say that it should be constructed from list_prefix ?

Author: I actually implemented `list()` as `list_prefix("")`, but it's also good to keep `list()` as a shortcut.
But we can't use `list()` to discover all the nodes in a hierarchy, since we need to prefix with `"meta/root/"`.

Protocol extensions
===================

Many types of extensions can exist for a Zarr protocol, they can be regrouped
in mostly two categories:

 - Core data type extensions – for example adding the ability to store fixed size
   types such as complex or datetime in chunks. These are directly declared in the
   array metadata ``data_type`` key.
 - Arrays extensions – non rectilinear grids, and variable length types.

> There are no group extensions as in Zarr v3.0.

Reviewer: ```suggestion
There are no group extensions in Zarr v3.0.
```
